ios/brave-ios/Sources/BraveVPN/Components/Region/BraveVPN+Region.swift

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import GuardianConnect
+import os.log
+
+extension BraveVPN {
+
+  /// Type of the vpn region
+  public enum RegionPrecision: Equatable {
+    case country, city
+  }
+
+  /// List of regions the VPN can connect to
+  /// This list is not static and it will be fetcghed with every app launch
+  static var allRegions: [GRDRegion] = []
+
+  /// Record last used region
+  /// It is used to hold details of the region when automatic selection is used
+  static var lastKnownRegion: GRDRegion?
+
+  /// Returns currently chosen region. Returns nil if automatic region is selected.
+  public static var selectedRegion: GRDRegion? {
+    helper.selectedRegion
+  }
+
+  /// Return the region last activated with the details
+  /// It will give region details for automatic
+  public static var activatedRegion: GRDRegion? {
+    helper.selectedRegion ?? lastKnownRegion
+  }
+
+  /// Switched to use an automatic region, region closest to user location.
+  public static func useAutomaticRegion() {
+    helper.select(nil)
+  }
+
+  /// Returns true whether automatic region is selected.
+  public static var isAutomaticRegion: Bool {
+    helper.selectedRegion == nil
+  }
+
+  public static func changeVPNRegion(to region: GRDRegion?, completion: @escaping ((Bool) -> Void))
+  {
+    if isConnected {
+      helper.disconnectVPN()
+    }
+
+    helper.select(region)
+
+    // The preferred tunnel has to be used for configuration
+    // Otherwise faulty configuration will be added while connecting
+    let activeTunnelProtocol = GRDTransportProtocol.getUserPreferredTransportProtocol()
+
+    helper.configureFirstTimeUser(for: activeTunnelProtocol, with: region) { success, error in
+      let subcredentials =
+        "Credentials \(GRDKeychain.getPasswordString(forAccount: kKeychainStr_SubscriberCredential) ?? "Empty")"
+
+      if success {
+        Logger.module.debug("Changed VPN region to \(region?.regionName ?? "default selection")")
+        completion(true)
+      } else {
+        Logger.module.debug("Connection failed: \(error ?? "nil")")
+        Logger.module.debug("Region change connection failed for subcredentials \(subcredentials)")
+        completion(false)
+      }
+    }
+  }
+
+  public static func populateRegionDataIfNecessary() {
+    helper.setPreferredRegionPrecision(kGRDRegionPrecisionCityByCountry)
+
+    helper.allRegions { regions, error in
+      guard let regions = regions else {
+        return
+      }
+
+      allRegions = regions
+    }
+  }
+
+  @MainActor
```

Comment:
Does this have to run on main?

Reply:
Nope not all fix at https://github.com/brave/brave-core/pull/24198/commits/db1d988f0db57643c18b4dd3cbca47d5c3adc98f